exchanges/orderbook/orderbook.go

```diff
@@ -206,62 +184,114 @@ func (b *Base) Update(bids, asks []Item) {
+// Asks should always go from a low price to a higher price
+func (b *Base) Verify() error {
+	if len(b.Asks) == 0 && len(b.Bids) == 0 {
+		return errNoOrderbook
 	}
-
-	lastPrice = 0
-	for x := range b.Asks {
-		if lastPrice != 0 && b.Asks[x].Price <= lastPrice {
-			sortAsks = true
-			break
+	for i := range b.Bids {
+		if b.Bids[i].Price == 0 {
+			return fmt.Errorf(bidLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errPriceNotSet)
+		}
+		if b.Bids[i].Amount == 0 {
+			return fmt.Errorf(bidLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errAmountNotSet)
+		}
+		if i != 0 {
+			if b.Bids[i].Price > b.Bids[i-1].Price {
+				return fmt.Errorf(bidLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errOutOfOrder)
+			}
+
+			if b.Bids[i].ID != 0 {
+				if b.Bids[i].ID == b.Bids[i-1].ID {
+					return fmt.Errorf(bidLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errIDDuplication)
+				}
+				continue
+			}
+
+			if !b.NotAggregated && b.Bids[i].Price == b.Bids[i-1].Price {
+				return fmt.Errorf(bidLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errDuplication)
+			}
 		}
-		lastPrice = b.Asks[x].Price
-	}
-
-	if sortBids {
-		sort.Sort(sort.Reverse(byOBPrice(b.Bids)))
 	}
 
-	if sortAsks {
-		sort.Sort((byOBPrice(b.Asks)))
+	for i := range b.Asks {
+		if b.Asks[i].Price == 0 {
+			return fmt.Errorf(askLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errPriceNotSet)
+		}
+		if b.Asks[i].Amount == 0 {
+			return fmt.Errorf(askLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errAmountNotSet)
+		}
+		if i != 0 {
+			if b.Asks[i].Price < b.Asks[i-1].Price {
+				return fmt.Errorf(askLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errOutOfOrder)
+			}
+
+			if b.Asks[i].ID != 0 {
+				if b.Asks[i].ID == b.Asks[i-1].ID {
+					return fmt.Errorf(askLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errIDDuplication)
+				}
+				continue
+			}
+
+			if !b.NotAggregated && b.Asks[i].Price == b.Asks[i-1].Price {
+				return fmt.Errorf(askLoadBookFailure, b.ExchangeName, b.Pair, b.AssetType, errDuplication)
+			}
+		}
 	}
+	return nil
 }
 
 // Process processes incoming orderbooks, creating or updating the orderbook
 // list
 func (b *Base) Process() error {
 	if b.ExchangeName == "" {
-		return errors.New(errExchangeNameUnset)
+		return errExchangeNameUnset
 	}
 
 	if b.Pair.IsEmpty() {
-		return errors.New(errPairNotSet)
+		return errPairNotSet
 	}
 
 	if b.AssetType.String() == "" {
-		return errors.New(errAssetTypeNotSet)
-	}
-
-	if len(b.Asks) == 0 && len(b.Bids) == 0 {
-		return errors.New(errNoOrderbook)
+		return errAssetTypeNotSet
 	}
 
 	if b.LastUpdated.IsZero() {
 		b.LastUpdated = time.Now()
 	}
 
-	b.Verify()
+	err := b.Verify()
+	if err != nil {
+		return err
+	}
 
 	return service.Update(b)
 }
+
+// Reverse reverses the order of orderbook items; some bid/asks are returned
+// without correct allignment. This is faster than using a sort algorithm
+func Reverse(elem []Item) {
```

Comment:
While appreciate this, reverse still does not guarantee that the order of bids and asks are correct given that an exchange can still send the bids and asks out of order. Verify helps, but returns an error upon failure rather than fixing any issues where a sort would actually guarantee its order

Reply:
Seeing as this is used for a small initial book/snapshot processing I am in slight agreeance; as for updates it's not really an issue. But it's a matter of reducing frictional cost and we need to take into account the probability of the books returned from endpoint as partial / snapshot diverting from inititial bid dsc/ ask asc (correct order bid/ask), dsc/dsc (correct bid, reverse ask) or asc/asc (correct ask, reverse bid) changing to random and vice versa due to API updates or database issues.